diff --git a/src/lib/tls/tls_channel.cpp b/src/lib/tls/tls_channel.cpp
index 37f3ec415..98edb618f 100644
--- a/src/lib/tls/tls_channel.cpp
+++ b/src/lib/tls/tls_channel.cpp
@@ -408,7 +408,7 @@ void Channel::process_handshake_ccs(const secure_vector<uint8_t>& record,
                                                          record_sequence);
                }
             }
-         else if(record_sequence == 0)
+         else/* if(record_sequence == 0)*/
             {
             create_handshake_state(record_version);
             }

index 37f3ec415..98edb618f 100644
--- a/src/lib/tls/tls_handshake_state.cpp
+++ b/src/lib/tls/tls_handshake_state.cpp
@@ -398,6 +398,9 @@ Handshake_State::choose_sig_format(const Private_Key& key,

       for(Signature_Scheme scheme : schemes)
          {
+         if(!for_client_auth && signature_scheme_is_known(scheme) == false)
+                  continue;
+
          if(signature_algorithm_of_scheme(scheme) == sig_algo)
             {
             if(std::find(allowed.begin(), allowed.end(), scheme) != allowed.end())
